crates/wasmtime/src/limits.rs

```diff
@@ -1,4 +1,108 @@
-pub use wasmtime_runtime::ResourceLimiter;
+/// Value returned by [`ResourceLimiter::instances`] default method
+pub const DEFAULT_INSTANCE_LIMIT: usize = 10000;
+/// Value returned by [`ResourceLimiter::tables`] default method
+pub const DEFAULT_TABLE_LIMIT: usize = 10000;
+/// Value returned by [`ResourceLimiter::memories`] default method
+pub const DEFAULT_MEMORY_LIMIT: usize = 10000;
+
+/// Used by hosts to limit resource consumption of instances.
+///
+/// An instance can be created with a resource limiter so that hosts can take into account
+/// non-WebAssembly resource usage to determine if a linear memory or table should grow.
+#[cfg_attr(feature = "async", async_trait::async_trait)]
+pub trait ResourceLimiter {
+    /// Notifies the resource limiter that an instance's linear memory has been
+    /// requested to grow.
+    ///
+    /// * `current` is the current size of the linear memory in bytes.
+    /// * `desired` is the desired size of the linear memory in bytes.
+    /// * `maximum` is either the linear memory's maximum or a maximum from an
+    ///   instance allocator, also in bytes. A value of `None`
+    ///   indicates that the linear memory is unbounded.
+    ///
+    /// This function should return `true` to indicate that the growing
+    /// operation is permitted or `false` if not permitted. Returning `true`
+    /// when a maximum has been exceeded will have no effect as the linear
+    /// memory will not grow.
+    ///
+    /// This function is not guaranteed to be invoked for all requests to
+    /// `memory.grow`. Requests where the allocation requested size doesn't fit
+    /// in `usize` or exceeds the memory's listed maximum size may not invoke
+    /// this method.
+    fn memory_growing(&mut self, current: usize, desired: usize, maximum: Option<usize>) -> bool;
+
+    /// Asynchronous version of `memory_growing`. This method is called for memory
+    /// grows on asynchronous stores, instead of `memory_growing`.
+    #[cfg(feature = "async")]
+    async fn memory_growing_async(
+        &mut self,
+        current: usize,
+        desired: usize,
+        maximum: Option<usize>,
+    ) -> bool;
```

Comment:
To prevent all consumers fo this trait from using `async_trait` I think it might be good to change this to have a default implementation of calling the `memory_growing` method?

I think it's a bit of a bummer to have both methods because if you're async you only want one and if you're not async you only want the other, but I don't know otherwise of a great way to manage this.

This does raise the question though of what happens when `Memory::new` (or `Memory::grow`) is called when async is configured? Even some async embeddings don't necessarily want async memory growth so enabling async doesn't guarantee that this method is configured. Ideally it's an embedder bug if `Memory::new` is called when async memory growth is desired though.

I think this might mean that we'll want a method here (perhaps default `false`?) which indicates which growth method to use (sync-or-async) and we dispatch with that on memory growth?

Reply:
So, I tried having a default implementation, but it turns out that with async_trait those arent object-safe and rustc rejects them.

The other possibility is to have two public traits, `ResourceLimiter` and `ResourceLimiterAsync`, and an internal `enum RLimiter { Sync(&mut dyn ResourceLimiter), Async(&mut dyn ResourceLimiterAsync) }` and `StoreInner::limiter` has type `Option<Box<dyn FnMut(&mut T) -> RLimiter>>` and some chrome on the config methods to wrap the closure's return value in the Sync or Async variant.

Reply:
Ah ok then yeah I like your idea if we can't have defaulted `async fn` in traits, that seems reasonable to me and also solves the issue of async-runtime-still-uses-sync-limiter.